feature_engine/datetime/datetime_subtraction.py

```diff
@@ -0,0 +1,350 @@
+                    "items in reference_variables * items in variables to "
+                    "combine. In other words, "
+                    "the transformer needs as many new variable names as reference "
+                    "variables to perform over the variables to "
+                    "combine."
+                )
+
+        if missing_values not in ["raise", "ignore"]:
+            raise ValueError("missing_values takes only values 'raise' or 'ignore'")
+
+        if not isinstance(drop_original, bool):
+            raise TypeError(
+                "drop_original takes only boolean values True and False. "
+                f"Got {drop_original} instead."
+            )
+
+        if not isinstance(dedupe_variable_pairs, bool):
+            raise TypeError(
+                "dedupe_variable_pairs takes only boolean values True and False. "
+                f"Got {dedupe_variable_pairs} instead."
+            )
+
+        self.reference_variables = reference_variables
+        self.variables_to_combine = variables_to_combine
+        self.output_unit = output_unit
+        self.dedupe_variable_pairs = dedupe_variable_pairs
+        self.new_variables_names = new_variables_names
+        self.missing_values = missing_values
+        self.drop_original = drop_original
+        self.dayfirst = dayfirst
+        self.yearfirst = yearfirst
+        self.utc = utc
+
+    def fit(self, X: pd.DataFrame, y: Optional[pd.Series] = None):
+        """
+        This transformer does not learn any parameter.
+
+        Parameters
+        ----------
+        X: pandas dataframe of shape = [n_samples, n_features]
+            The training input samples. Can be the entire dataframe, not just the
+            variables to transform.
+
+        y: pandas Series, or np.array. Default=None.
+            It is not needed in this transformer. You can pass y or None.
+        """
+
+        # check input dataframe
+        X = _is_dataframe(X)
+
+        # find or check for datetime variables to combine
+        self.variables_to_combine_ = _find_or_check_datetime_variables(
+            X, self.variables_to_combine
+        )
+
+        # find or check for datetime reference variables
+        self.reference_variables_ = _find_or_check_datetime_variables(
+            X, self.reference_variables
+        )
+
+        # check if dataset contains na
+        if self.missing_values == "raise":
+            _check_contains_na(X, self.reference_variables_)
+            _check_contains_na(X, self.variables_to_combine_)
+
+            _check_contains_inf(X, self.reference_variables_)
+            _check_contains_inf(X, self.variables_to_combine_)
+
+        variable_pairs = list(itertools.product(self.reference_variables_,
+                                                self.variables_to_combine_))
+
+        if self.dedupe_variable_pairs:
+            # remove the pairs consisting of the same elements
+            # then sort the tuple values and dedupe them
```

Comment:
so let's say we have 
```
variable_pairs = [('v1', 'v2'), ('v2', 'v1')]
```
the tuple sorting recognizes that these features will contain "essentially" the same info and retains only 1 of them: [('v1', 'v2')]

Reply:
the idea of separating into variables to combine and reference variables is that we substract the second from the first. And do not make all possible combinations.

If the user wants to do v1 -v2 and then v2-v1 they would have to specifically say so by passing v1 and v2 to both the ref and the to combine params. Otherwise, this would not happen.

Which is probably a good point, because we would not use itertools to create all possible combinations?

Reply:
I followed the convention used in DatetimeFeatures, where the variable parameters are optional. When the reference and combine variables are not specified, _find_or_check_datetime_variables selects all the possible datetime variables.

This supports what would be my primary use case for the class, which is to create all of the possible datetime differences as features.

However, the main issue w/ this approach is that the numbers of subtraction features grow rapidly/exponentially as you increase the number datetime variables. Hence, this is why I proposed using the optional dedupe_variable_pairs parameter so that the class doesn't explode the size of the data.

What do you think?

Reply:
I see. 

For this transformer, I would follow the convention of the CombineWithReferenceFeature, where the parameters `variables_to_combine` and `reference` can not be left to None. The user needs to pass the variables for it to work.

The idea of the transformers in the creation folder is that the user creates features based on some **domain knowledge**, and not combine features for the sake of combining features.

We have a different issue for creating all possible combinations #84 which we have not prioritized because of the problem of expansion of the feature space, and also because so far, Feature-engine has mostly focused on creating features that can be used in models that serve people, and thus need to be interpreted.

So my idea for this transformer is that given a bunch of datetime variables, the user wants to create elapsed time by combining those with 1 or more reference ones. In this context, variables_to_combine could accept the None argument and then select all datetime variables. But reference would have to be enforced.

Then, we would not create all possible combinations, just a-b where b is reference. 

The deduped functionality could still be necessary, because if a user passes variables_to_combine = [var_a, var_b_,var_c] and ref=[var_c, var_d]), we could still eliminate the pair varc-varc. Or not: we could also choose not to do this and then force the user to use 2 transformers, which is probably cleaner and more readable from the user's perspective. So maybe the deduped funtionality is also something to debate.

-  variables_to_combine = [var_a, var_b_],  ref=[var_c])
-  variables_to_combine = [var_a, var_b_,var_c],  ref=[var_d])

to obtain the desired result.

If you think combining all datetime with all datetime would be of use, then let's make that transformer in a different PR with a different name and structure. But first, let's create an issue so we can have some discussions around it.

I would start simple, let's first draft a transformer that looks pretty much like the combine with feature reference, let's write some tests, and once we are happy with the basic functionality, let's see if we also want to dedupe or not.

thoughts?